src/libraries/System.Private.CoreLib/src/System/Convert.cs

```diff
@@ -2960,6 +2960,65 @@ public static byte[] FromHexString(ReadOnlySpan<char> chars)
             return result;
         }
 
+        /// <summary>
+        /// Converts the string, which encodes binary data as hex characters, to an equivalent 8-bit unsigned integer span.
+        /// </summary>
+        /// <param name="source">The string to convert.</param>
+        /// <param name="destination">
+        /// The span in which to write the converted 8-bit unsigned integers. When this method returns false,
+        /// either the span remains unmodified or contains an incomplete conversion of <paramref name="source"/>,
+        /// up to the last valid character.
+        /// </param>
+        /// <param name="bytesWritten">When this method returns, contains the number of bytes that were written in <paramref name="destination"/>.</param>
+        /// <returns>true if the conversion was successful; otherwise, false.</returns>
+        /// <exception cref="ArgumentNullException">Passed string <paramref name="source"/> is null.</exception>
+        public static bool TryFromHexString(string source, Span<byte> destination, out int bytesWritten)
+        {
+            ArgumentNullException.ThrowIfNull(source);
+
+            return TryFromHexString(source.AsSpan(), destination, out bytesWritten);
+        }
+
+        /// <summary>
+        /// Converts the span of chars, which encodes binary data as hex characters, to an equivalent 8-bit unsigned integer span.
+        /// </summary>
+        /// <param name="source">The span to convert.</param>
+        /// <param name="destination">
+        /// The span in which to write the converted 8-bit unsigned integers. When this method returns false,
+        /// either the span remains unmodified or contains an incomplete conversion of <paramref name="source"/>,
+        /// up to the last valid character.
+        /// </param>
+        /// <param name="bytesWritten">When this method returns, contains the number of bytes that were written in <paramref name="destination"/>.</param>
+        /// <returns>true if the conversion was successful; otherwise, false.</returns>
+        public static bool TryFromHexString(ReadOnlySpan<char> source, Span<byte> destination, out int bytesWritten)
+        {
+            var length = source.Length;
```

Comment:
```suggestion
            int length = source.Length;
```
Use concrete types here. Same on other places.

Perf-wise this could be `nuint`, so all the division and modulo operations are treated as (fast) bit-operations.

Reply:
> Perf-wise this could be nuint, so all the division and modulo operations are treated as (fast) bit-operations.

That's interesting. We rarely use native width types except in interop contexts. @jkotas @EgorBo is it correct that in hot code of this type, it can have perf benefit for simple storage of counters and offsets?

Reply:
Afair, JIT knows that Span.Length (and Array.Length as well) are never negative so it doesn't need extra efforts from the C# side 🙂 e.g.:
```cpp
int Foo(Span<int> s) => s.Length % 4;
```
```asm
; Method Foo(System.Span`1[int]):int:this
       mov      eax, dword ptr [rdx+08H]
       and      eax, 3
       ret      
; Total bytes of code: 7
```
(some of the changes landed in .NET 8.0 so might be visible on sharplab yet)

Reply:
Right -- I'm not aware of patterns where nuint gives better code.

Reply:
> JIT knows that Span.Length (and Array.Length as well) are never negative

Thanks for the reminder -- I'll forget sometimes about the new JIT-goodness.

> We rarely use native width types except in interop contexts

Quite a lot of vectorized code uses the native width types, because memory operations don't need (sign-) extending moves then (e.g. `Unsafe.Add(ref ptr, intVariable)` vs. `Unsafe.Add(ref ptr, nuintVariable)`.

Reply:
Would `Math.DivRem` be more efficient here? Especially once it becomes an [intrinsic](https://github.com/dotnet/runtime/issues/82194).

Reply:
> Would `Math.DivRem` be more efficient here? Especially once it becomes an [intrinsic](https://github.com/dotnet/runtime/issues/82194).

I doubt it, since the JIT optimizes `length % 2 != 0` to `(length & 1) != 0`.

Reply:
> the division done a few lines later

The JIT can optimize `(uint)length / 2` to `(uint)length >> 1`. On X86, `DivRem` would likely be implemented using a `div` instruction, which is an relatively expensive operation.